lib/ret_web/channels/hub_channel.ex

```diff
@@ -343,35 +365,57 @@ defmodule RetWeb.HubChannel do
     socket.assigns |> Map.take([:presence, :profile, :context])
   end
 
-  defp join_with_hub(%Hub{entry_mode: :deny}, _socket, _endpoint, _auth_token) do
+  defp join_with_hub(%Hub{entry_mode: :deny}, _account, _socket, _params) do
     {:error, %{message: "Hub no longer accessible", reason: "closed"}}
   end
 
-  defp join_with_hub(%Hub{} = hub, socket, push_subscription_endpoint, auth_token) do
+  defp join_with_hub(%Hub{} = hub, %Account{}, _socket, %{
+         hub_requires_oauth: true,
+         account_has_provider_for_hub: false
+       }),
+       do: require_oauth(hub)
+
+  defp join_with_hub(%Hub{}, %Account{}, _socket, %{
+         hub_requires_oauth: true,
+         account_has_provider_for_hub: true,
+         account_can_join: false
+       }),
+       do: deny_join()
+
+  defp join_with_hub(%Hub{} = hub, nil = _account, _socket, %{
+         hub_requires_oauth: true,
+         has_valid_bot_access_key: false,
+         has_perms_token: false
+       }),
+       do: require_oauth(hub)
+
+  defp join_with_hub(%Hub{}, nil = _account, _socket, %{
+         hub_requires_oauth: true,
+         has_valid_bot_access_key: false,
+         has_perms_token: true,
+         perms_token_can_join: false
```

Comment:
maybe increase deny scope to anytime `perms_token_can_join` or `account_can_join` is false?

Reply:
The two are mutually exclusive, so I don't think that works, unless I misunderstood you.

Reply:
i meant instead of matching on all the other fields, just write two overloads -- one that maches on `perms_token_can_join: false` and the other on `account_can_join: false` -- it seems that no matter what other fields on that data structure the caller passes in, if those flags are false we want to deny access.

Reply:
Yeah, that won't work since `account_can_join` is always false when `perms_token_can_join`, and vice-versa. Pretty sure I have the minimal set of required fields already.